pkg/objstore/s3/s3.go

```diff
@@ -0,0 +1,148 @@
+// Package s3 implements common object storage abstractions against s3-compatible APIs.
+package s3
+
+import (
+	"context"
+	"io"
+	"strings"
+
+	"github.com/minio/minio-go"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	opObjectsList  = "ListBucket"
+	opObjectInsert = "PutObject"
+	opObjectGet    = "GetObject"
+	opObjectStat   = "StatObject"
+	opObjectDelete = "DeleteObject"
+)
+
+// DirDelim is the delimiter used to model a directory structure in an object store bucket.
+const DirDelim = "/"
+
+// Bucket implements the store.Bucket interface against s3-compatible APIs.
+type Bucket struct {
+	bucket   string
+	client   *minio.Core
+	opsTotal *prometheus.CounterVec
+}
+
+// Config encapsulates the necessary config values to instantiate an s3 client.
+type Config struct {
+	Bucket    *string
+	Endpoint  *string
+	AccessKey *string
+	SecretKey *string
+	Insecure  *bool
+}
+
+// Validate checks to see if any of the s3 config options are set.
+func (conf *Config) Validate() bool {
+	if *conf.Bucket == "" &&
```

Comment:
Probably should error as soon as the bucket is the empty string. Returning an error instead of a boolean would also be preferable.

Reply:
Agreed, will fix.